File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int max_val = 0;
    int trial(TreeNode* root){
        if(root == NULL){
            return 0;
        }

        int current_node = trial(root->left) + trial(root->right);
        int carried_val = max(trial(root->left) ,trial(root->right));

        if(current_node>max_val){
            max_val = current_node;
        }

        return 1+carried_val;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        trial(root);
        return max_val;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        length = len(heights)
        left = 0
        right = length - 1
        area = 0
        while right>left:
            product = (right - left)*min(heights[left],heights[right])
            area  = max(area, product)
            if(heights[left]<heights[right]):
                left+=1
                continue

            if(heights[right]<=heights[left]):
                right-=1
                continue

        return area
        